Objetivo 8

Balrrach opened the pull request:
* [x] ¿Has justificado tu elección de un framework REST y explicado por qué no has
   elegido ni `flask` ni `express`?
* [x] ¿Las rutas están organizadas alrededor de un recurso?

Balrrach committed:
Clean

Es conveniente mantener las discusiones de todas las
herramientas en un mismo fichero

Balrrach committed:
.

Balrrach committed:
Feat #30

Es necesario justificar la elección de la librería de logging
para poder revisarla en un futuro y porque la justificación
fomenta la búsque da una librería que se adecúe a los requisitos
del sistema.

Balrrach committed:
Fix

Se pretende que los tests solo corran cuando terminen las GHA
encargadas de crear el contenedor o de actualizarlo.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Feat #30

Se ha reconsiderado la elección puesto que ulog no tiene ssoporte
nativo para typescript.

Balrrach committed:
Clean

Balrrach committed:
Feat #30

Ni los logs ni los archivos de configuración pueden terminar
en el repositorio.

Balrrach committed:
Feat #30

Las decisiones técnicas relacionadas con las librerías
utilizadas para la configuración remota y para la carga de las
variables de entorno han de estar justificadas para su futura
revisión.

Balrrach committed:
Clean

Balrrach committed:
Clean

Balrrach committed:
Clean

Balrrach committed:
.

Balrrach committed:
Clean

El mensaje es más claro

Balrrach committed:
Feat #30

El orden en el que se toman las configuraciones es:
1. Servidor
2. Enviroment
3. Valores predeterminados
De esta forma la aplicación siempre tiene una configuración
preparada para utilizar.

Puesto que no se puede devolver una promesa de un objeto dada
su propiedad inherente de *disponibilidad(Readiness)*, la
solución que se ha tomado es crear una atributo que corresponde
a una promesa y ejecutar los metodos necesarios cuando la
promesa se verifique. La solución detallada se puede encontrar
en: https://stackoverflow.com/questions/35743426/async-constructor-functions-in-typescript

Balrrach committed:
Clean

Tiene más sentido tener la ruta en primer lugar puesto que conforma la
primera parte de la ruta.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Clean

Balrrach committed:
Fix

No se están utilizando los mismos tags en run_test que en
create_pr_container puesto que no se están usando los mismos
eventos(on: pull request ===vs=== on: push)

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Feat #30, #33

Es conveniente utilizar una clase controladora e instanciarla
al inicio del programa asegurándose de esta manerar que todas sus
dependencias están correctamente inicializadas desde el primer momento.

Balrrach committed:
Feat #30, #33

La clase controladora se exporta para su uso en el resto de clases.
Se implementa getLogger() puesto que es necesario acceder al logger.

Balrrach committed:
Clean

Balrrach committed:
Feat #30, #33

Primer intento de implentacion de un controlador asincrono.
No funciona correctamente porque, por un motivo desconocido, no se está
esperando a que la instancia config del controlador se termine de
instanciar antes de utilizarla en el index o en los test aunque el test
específico de la clase configuración funciona cocorrectamente.

Balrrach committed:
.

Balrrach committed:
Fix

No se puede mantener la estructura de un contenedor por PR porque JJ
verifica sus tests contra el contenedor con tag latest

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
.

Balrrach committed:
Feat #33

La dependencia no es del método sino de estado de disponibilidad del
objeto.

Balrrach committed:
Clean

No es neceasario esperar a que el objeto de configuración esté preparado
en index puesto que no se hace un uso explicito del controlador.

Balrrach committed:
Closes #34

El problema era consecuencia de que el código que encerraba la promesa
de la clase Config tiraba las llamadas asíncronas e inmediatamente
resolvía a undefined. Como consecuencia el estado Ready no resolvía el
problema. La solución ha sido considerar Ready como el conjunto de
promesas con las que realmente está ligado. Más información en:
https://stackoverflow.com/questions/70501382/related-promises-in-object-creation

Balrrach committed:
Feat #34

De misma forma ocurre con el Controlador. El logger no tenía por qué
estar instanciado lo estaba por el mero hecho de que el objeto de
configuración se instanciaba automáticamente por error. Al resolver su
asincronía el esto ya no tiene que ser así y hay que asegurarse que el
logger esta correcamente instanciado.

No termino de entender porque estaba corriendo correctamente antes de
este arreglo.

Balrrach committed:
.

Balrrach committed:
Fix: Typo

Balrrach committed:
Fix

El campo configuración tiene que hacer referencia al fichero que maneja
la configuración no al fichero que la contiene(que ni siquiera se sube
al repositorio).

Balrrach committed:
Fix: Typo

Balrrach committed:
Feat #36

La decisión del framework utilizado para la implementación del REST API
ha de estar documentada de forma que sea revisable en un futuro en caso
de necesidad.

Balrrach commented:
El PR realmene comienza en 5831597. El resto de issues son del PR correspondiente al objetivo 7 que no estaba mergeado cuando se creó este último.

Balrrach committed:
Clean

Lo mejor es que todas las variables tenga el tipo más cerrado posible
para aprovechar al máximo las características del tipado fuerte.

Balrrach committed:
Merge branch 'Objetivo-7' into Objetivo-8

Balrrach committed:
.

Balrrach committed:
Feat #30, #33

Se ha decidido transformar la creación del configurador en síncrona
porque en caso contrario todas las llamadas al mismo(en esencia todo el
programa porque es necesario para instanciar el logger, el servidor...)
se convertían asíncronas haciendo su gestión terriblemente tediosa para
las supuestas ventajas que podía tener la asincronía.

Balrrach committed:
Merge branch 'Objetivo-7' into Objetivo-8

Balrrach committed:
Feat #36

La configuración del servidor ha de incluirse en el objeto de
configuración para ser posteriormente leída por el Controlador que será
el encargado de instanciar el servidor.

Balrrach committed:
Feat #36

La implementación de la API correctamente documentada en el archivo del
proyecto.

Balrrach committed:
Fix #36

Los verbos de creación y actualización estaban intercambiados.

Balrrach committed:
Feat #36

La justificación de los recursos que se acceden con la API, que
grado de acceso tiene cada uno y cómo se realiza este acceso ha de estar
documentada para que los desarrolladores puedan hacer un uso adecuado de
la API.

Balrrach committed:
Feat #30, #33

Es conveniente añadir un archivo de configuracion de prueba al proyecto
para correr los test de la clase de configuración y para que los desarrolladores
tengan una guía para la creación de su propio archivo de configuracón.

Balrrach committed:
Feat #30, #33

Es conveniente añadir un archivo de configuracion de prueba al proyecto
para correr los test y para que los desarrolladores tengan una guí

Balrrach committed:
Merge branch 'Objetivo-7' of github.com:Balrrach/IV-Proyecto into Objetivo-7

Balrrach committed:
Fix

Se arregla el gitignore para incluir el archivo de configuración de
prueba.

Balrrach committed:
Merge branch 'Objetivo-7' into Objetivo-8

Balrrach committed:
.

Balrrach committed:
Fix

Se vuelve a instalar hapi porque la rama no lo tenía instalado.

Balrrach committed:
Clean

Balrrach committed:
Fix

Balrrach committed:
Clean

Balrrach commented:
Revisión @JJ.

Balrrach commented:
No le he hecho el pr al repositorio porque no tengo el Objetivo-7 superado

Balrrach committed:
Clean

Se añade la posibilidad de limpiar el directorio de todos los archivos
generados.

Balrrach committed:
Fix

JJ commented:
> No le he hecho el pr al repositorio porque no tengo el Objetivo-7 superado y fallarían los test. Igualmente cuando corrijas, por favor, échale un ojo porque es independiente del Objetivo-7 y necesito saber cómo va para seguir avanzando.

Puedes hacerlo inicialmente como un PR a la rama del objetivo 7,  y cuando lo fusiones simplemente cambias la rama a la que vas a hacer PR.

Balrrach committed:
Merge branch 'main' into Objetivo-8

Balrrach committed:
Fix

Balrrach committed:
Clean

Balrrach committed:
Fix

No tiene sentido que las clases que implementan un patrón singletón
tengan atributos en su constructor.

Balrrach committed:
Feat #30

Las clases no han de instanciar su propia creación por el principio
open/close y por *separation of concerns*.

Balrrach committed:
Clean

Balrrach committed:
.

JJ commented:
⛹ Revisores → @modejota @Slowmybrosh

Balrrach commented:
Revisión @JJ.

modejota commented on docs/tools.md:
```diff
@@ -90,3 +90,19 @@ El primero es que el estado de mantenimiento de dotenv es mucho mejor que el de
+Se han considerado los siguientes frameworks: [nest](https://www.npmjs.com/package/@nestjs/core), [fastify](https://www.npmjs.com/package/fastify), [hapi](https://www.npmjs.com/package/@hapi/hapi), [restify](https://www.npmjs.com/package/restify).
+
+- Nest es la alternativa más popular. Nest es un proyecto apoyado por grandes multinacionales como Redhat y empresas como NX y Labster que tiene como objetivo resolver los problemas arquitectónicos frecuentes en las aplicaciones escritas en javascript/typescript: "... However, while plenty of superb libraries, helpers, and tools exist for Node (and server-side JavaScript), none of them effectivel ...
+- Fastify es la segunda alternativa más popular y la más novedosa. Esta librería esta centrada en la eficiencia siendo esta uno de sus principales atractivos; es la alternativa más eficiente de las consideradas según sus propios benchmark que se pueden encontrar en https://github.com/fastify/benchmarks. Su estado de mantenimiento es muy bueno y se integra bien con el logger elegido pero se integr ...
```
Sorry por no haber comentado nada, cosas de tomarse el día 1 de fiesta. Como ya lo has puesto para revisión, solo te diré como comentario que fastify tiene la funcion "inject" para hacer peticiones HTTP "fake" en los test, con JEST funciona bastante bien, ya no sé con Mocha, pero supongo será 3/4 de lo mismo.

Balrrach replied:
> Sorry por no haber comentado nada, cosas de tomarse el día 1 de fiesta. Como ya lo has puesto para revisión, solo te diré como comentario que fastify tiene la funcion "inject" para hacer peticiones HTTP "fake" en los test, con JEST funciona bastante bien, ya no sé con Mocha, pero supongo será 3/4 de lo mismo.

Aunque lo haya puesto para revisión podéis comentar lo que queráis, si me da tiempo a cambiarlo antes de la revisión, lo haré. Tienes razón con el tema de los injects, lo que realmente no me gusta es que la declaración de las rutas y sus llamadas sean distintas(entre otras cosas). Voy a actualizar la documentación.

Balrrach replied:
Hecho 6db2eba.

Balrrach committed:
Feat #36

Era necesario completar la documentación acerca de la elección de la
librería en la que se apoya la implementación de la REST API porque no
estaban claros los motivos de preferencia respecto de la librería escogida
frente a fastify.

JJ commented on docs/api.md:
```diff
@@ -0,0 +1,62 @@
+| GET	| /product	| Obtención de todos los productos			|
+| GET	| /product/id	| Obtención del producto relacionado con la id		|
+| POST	| /product	| Creación de un nuevo producto				|
+| PUT	| /product	| Actualización de un producto				|
```
PUT siempre debe usar un URI, porque es idempotente.

JJ replied:
Lo que quiere decir que no cambia el estado.

Balrrach replied:
Te estás refiriendo a la documentación que no está actualizada porque se me olvidó. El verdadero diseño de la API está reflejado en el código que siempre está actualizado.

JJ replied:
Me estoy refiriendo a lo que veo, claro. En este objetivo se trataba de hacer sólo la documentación.

Balrrach committed:
Fix #40

Bien es cierto, PUT siempre se realiza sobre un URI.

JJ commented on docs/tools.md:
```diff
@@ -90,3 +90,24 @@ El primero es que el estado de mantenimiento de dotenv es mucho mejor que el de
+- Nest es la alternativa más popular. Nest es un proyecto apoyado por grandes multinacionales como Redhat y empresas como NX y Labster que tiene como objetivo resolver los problemas arquitectónicos frecuentes en las aplicaciones escritas en javascript/typescript: "... However, while plenty of superb libraries, helpers, and tools exist for Node (and server-side JavaScript), none of them effectivel ...
+- Fastify es la sugunda alternativa más popular y la más novedosa. Esta librería esta centrada en la eficiencia siendo esta uno de sus principales atractivos; es la alternativa más eficiente de las consideradas según sus propios benchmark que se pueden encontrar en https://github.com/fastify/benchmarks. Su estado de mantenimiento es muy bueno y se integra bien con el logger y la librería de test  ...
+- Restify es la tercera de las alternativas consideradas. Es utilizada por grandes empresas/organizaciones como npm, netflix, pinterest o napster. La librería fue descartada porque, a pesar de cumplir todos los requisitos de la lista tiene un estado de conservación relativamente malo(76 issues de los cuales los últimos no han sido tan siquiera respondidos) lo cual convierte su elección en un tant ...
+- Por último está hapi que, a pesar de ser la librería menos popular, es la elegida. Hapi se desarrollo gestionar Walmart durante el Black Friday y es utilizada por grandes empresas como Brave(web browser) y Beats. Verifica todos los requisitos necesarios y convenientes y, además, tiene un estado de conservación bueno y una documentación muy clara y concisa que facilitan enormemente su uso.
```
Podías poner enlaces o algo y al final qué es lo que has elegido. O resaltar más que has elegido esta. No dices nada de los criterios, y dices cosas que no vienen a cuento como que la usa Brave (¿¿¿???)

Balrrach replied:
> Podías poner enlaces o algo y al final qué es lo que has elegido.

Todos los framework están enlazados a sus correspondientes páginas la primera vez que son mencionados y se indica claramente el framework escogido en el instante en el que se realiza su justificación.

> No dices nada de los criterios

En algunos casos se mencionan de forma implícita y en otros de forma explícita y toda la justificación gira en torno a los mismos:
- Nest: Esto implica que las herramientas necesarias para construir el API están pensadas para usarse con el resto de las herramientas del proyecto pero sobre todo que hay que seguir la estructura de aplicación propuesta. Por lo tanto no verifica la tercera condición y ha sido descartado.

- Fastify: Su estado de mantenimiento es muy bueno y se integra bien con el logger y la librería de test escogidos. Por otra parte no verifica ninguna los requisitos convenientes...

- Restify: La librería fue descartada porque, a pesar de cumplir todos los requisitos de la lista tiene un estado de conservación relativamente malo.

- Hapi: Verifica todos los requisitos necesarios y convenientes y, además, tiene un estado de conservación bueno y una documentación muy clara y concisa...

> y dices cosas que no vienen a cuento como que la usa Brave (¿¿¿???)

Sólo me pareció un dato curioso, evidentemente no forma parte de la justificación pero el texto no es sólo justificación, también es documentación.

JJ replied:
El problema es "la forma implícita". Hay que decodificar lo que dice para ver si cumplen los criterios o no. En cuanto a lo de Brave, "Datos curiosos" pueden ser interesantes en algún contexto, para el que tiene que evaluar este PR sólo es ruido.

JJ commented on docs/tools.md:
```diff
@@ -90,3 +90,24 @@ El primero es que el estado de mantenimiento de dotenv es mucho mejor que el de
+
+Se han considerado los siguientes frameworks: [nest](https://www.npmjs.com/package/@nestjs/core), [fastify](https://www.npmjs.com/package/fastify), [hapi](https://www.npmjs.com/package/@hapi/hapi), [restify](https://www.npmjs.com/package/restify).
+
+- Nest es la alternativa más popular. Nest es un proyecto apoyado por grandes multinacionales como Redhat y empresas como NX y Labster que tiene como objetivo resolver los problemas arquitectónicos frecuentes en las aplicaciones escritas en javascript/typescript: "... However, while plenty of superb libraries, helpers, and tools exist for Node (and server-side JavaScript), none of them effectivel ...
```
¿Es flexible, integrable, eficiente y tiene soporte para Typescript? Después de establecer los criterios ni siquiera los mencionas.

Balrrach replied:
Ya he respondido a esto arriba...

JJ replied:
Sí, pero el que te lo pregunte indica que no está nada claro. ¿Qué trabajo os cuesta poner una tabla con vuestros criterios y si se cumplen o no?

JJ replied:
La frase en inglés está copiada aparentemente de https://www.mitzon.com/mooc/develop-restful-web-services-using-nestjs-and-mongodb/ y ni siquiera tiene sentido. Es una web con mucho *SEO juice* que atrae a los incautos. ¿Qué es la "Arquitectura -"?

*Use the source, Luke*: Si te vas a [ella](https://docs.nestjs.com/v4/), lo que dice es:
> none of them effectively solve the main problem - the architecture.

Lo que sí tiene sentido (y es también irrelevante a su elección o no como framework en este contexto)

JJ commented on docs/tools.md:
```diff
@@ -90,3 +90,24 @@ El primero es que el estado de mantenimiento de dotenv es mucho mejor que el de
+Se han considerado los siguientes frameworks: [nest](https://www.npmjs.com/package/@nestjs/core), [fastify](https://www.npmjs.com/package/fastify), [hapi](https://www.npmjs.com/package/@hapi/hapi), [restify](https://www.npmjs.com/package/restify).
+
+- Nest es la alternativa más popular. Nest es un proyecto apoyado por grandes multinacionales como Redhat y empresas como NX y Labster que tiene como objetivo resolver los problemas arquitectónicos frecuentes en las aplicaciones escritas en javascript/typescript: "... However, while plenty of superb libraries, helpers, and tools exist for Node (and server-side JavaScript), none of them effectivel ...
+- Fastify es la sugunda alternativa más popular y la más novedosa. Esta librería esta centrada en la eficiencia siendo esta uno de sus principales atractivos; es la alternativa más eficiente de las consideradas según sus propios benchmark que se pueden encontrar en https://github.com/fastify/benchmarks. Su estado de mantenimiento es muy bueno y se integra bien con el logger y la librería de test  ...
```
typo - plugin

JJ commented on package.json:
```diff
@@ -18,15 +19,17 @@
 	},
 	"homepage": "https://github.com/Balrrach/IV-Proyecto#readme",
 	"devDependencies": {
+		"@hapi/hapi": "^20.2.1",
```
Pero ¿cómo va a ser esto devDepencency? Tendrá que ser de producción. Dev son las dependencias que sólo se instalan durante el desarrollo, test y  cosas así. La mayoría de las que hay debajo, imagino (salvo lo relacionado con hapi)

Balrrach replied:
Cierto, me equivoqué al instalarla, es una dependencia de producción.

JJ reviewed (changes requested):
Dado que la justificación no es demasiado buena, y que además el interfaz no está demasiado bien diseñado, estaba en el límite de ser aceptado. Sin embargo, lo que lo alejó del límite es que hayas instalado una dependencia que es claramente de producción como dependencia de desarrollo. Lo siento, pero no pasa.

Balrrach commented:
> Dado que la justificación no es demasiado buena

¿Por qué?

>y que además el interfaz no está demasiado bien diseñado

¿Por qué?

JJ commented:
> > Dado que la justificación no es demasiado buena
> 
> ¿Por qué?


Porque, como tú has dicho, es "implícita" y no explícita e incluso explícitamente no queda claro qué es lo que cada uno de los frameworks cumplen y qué no. Añades detalles irrelevantes en la descripción de los mismos, y eventualmente no se cumple el objetivo; que se pueda seguir claramente el proceso de aprendizaje de los conceptos de este objetivo y tu toma de decisiones.
> 
> > y que además el interfaz no está demasiado bien diseñado
> 
> ¿Por qué?

Uso de PUT de manera no idempotente (o no actualización de la documentación, que viene a ser lo mismo, porque no hay forma de evaluarlo que no sea la documentación)

En todo caso, como te he comentado, eso no habría impedido que pasaras si hubieras añadido la dependencia correctamente.

Balrrach commented:
> Porque, como tú has dicho, es "implícita" y no explícita e incluso explícitamente no queda claro qué es lo que cada uno de los frameworks cumplen y qué no.

Nest no verifica 3, Fastify no verifica ninguno de los requisitos convenientes, Restify, pese a verificar todos los requisitos no tiene un buen estado de mantenimiento y Hipo verifica todos los requisitos y está correctamente mantenido y documentado. ¿De verdad todo esto no queda claro en el documento?.
 
> Uso de PUT de manera no idempotente (o no actualización de la documentación, que viene a ser lo mismo, porque no hay forma de evaluarlo que no sea la documentación)

Si no hay otra forma de evaluarlo que no sea la documentación no entiendo como fueron aprobados estos dos objetivos antes del mío: https://github.com/modejota/StoragIV/pull/35, https://github.com/Slowmybrosh/DietApp/pull/47. Salvo que esté muy equivocado ninguno de ellos tenía documentada la estructura de la API y, sin embargo, superaron el objetivo porque se entendió que quedaba reflejada en el propio código. Parece que, si ese es el caso, se me está penalizando por haber documentado la API puesto que, no sólo es más trabajo sino que, además, al olvidárseme actualizarla el objetivo se considera no superado cuando realmente el fallo no existe en el código.

JJ commented:
> > Porque, como tú has dicho, es "implícita" y no explícita e incluso explícitamente no queda claro qué es lo que cada uno de los frameworks cumplen y qué no.
> 
> Nest no verifica 3, Fastify no verifica ninguno de los requisitos convenientes, Restify, pese a verificar todos los requisitos no tiene un buen estado de mantenimiento y Hipo verifica todos los requisitos y está correctamente mantenido y documentado. ¿De verdad todo esto no queda claro en el documento?.

Tú mismo has dicho que está "implícito".   Y no, te estoy diciendo que no queda claro. Vamos a ver por qué no queda claro. Y de veras que me parece un poco increíble que tenga que poner yo esto aquí, pero es tu derecho a la reclamación y yo lo hago con gusto. Quedamos en que había 4 condiciones: flexible, eficiente, integrable y soporte para tipos de typescript (no considero los "convenientes"). Vamos a ver el primero de ellos:

> Nest es la alternativa más popular.

Irrelevante

> Nest es un proyecto apoyado por grandes multinacionales como Redhat y empresas como NX y Labster que tiene como objetivo resolver los problemas arquitectónicos frecuentes en las aplicaciones escritas en javascript/typescript: "... However, while plenty of superb libraries, helpers, and tools exist for Node (and server-side JavaScript), none of them effectively solve the main problem of - Architecture". 

Irrelevante.

> Esto implica que las herramientas necesarias para construir el API están pensadas para usarse con el resto de las herramientas del proyecto pero sobre todo que hay que seguir la estructura de aplicación propuesta. Por lo tanto no verifica la tercera condición y ha sido descartado.

La tercera condición es "integrable con test y logging". ¿No es integrable con test y logging? ¿Es esto tan importante? ¿"Por lo tanto" de que tenga su propio generador de aplicaciones se deduce que no se puedan usar herramientas de test y logging?

¿Ves lo que pasa si no pones unas casillas y marcas o no las condiciones que cumples? Es más, esta es una nueva condición: que te permita integrarse a un proyecto existente (lo que, por cierto, debería ser un prerrequisito, porque el proyecto *ya* existe), que no has mencionado como requisito.

Si quieres sigo con el resto, pero es que en casi todos los casos es así. Insisto, tú has dicho que es "implícita". Si quieres cambio la corrección "débil" por "implícita", pero creo que ya habrás cogido la idea.

> 
> > Uso de PUT de manera no idempotente (o no actualización de la documentación, que viene a ser lo mismo, porque no hay forma de evaluarlo que no sea la documentación)
> 
> Si no hay otra forma de evaluarlo que no sea la documentación no entiendo como fueron aprobados estos dos objetivos antes del mío: [modejota/StoragIV#35](https://github.com/modejota/StoragIV/pull/35), [Slowmybrosh/DietApp#47](https://github.com/Slowmybrosh/DietApp/pull/47). Salvo que esté muy equivocado ninguno de ellos tenía documentada la estructura de la API y, sin embargo, superaron el objetivo porque se entendió que quedaba reflejada en el propio código. Parece que, si ese es el caso, se me está penalizando por haber documentado la API puesto que, no sólo es más trabajo sino que, además, al olvidárseme actualizarla el objetivo se considera no superado cuando realmente el fallo no existe en el código.

No, existe en un issue que te abrí para ayudarte a superar el objetivo y que simplemente obviaste. Ayuda que, por cierto, no tuvieron tus compañeros.

Vamos a ver. ¿Estamos de acuerdo que has cometido un error en la documentación? ¿Un error que te señalé en #40 y que no arreglaste *en la documentación*, provocando una discrepancia entre esta y el código que es tan errónea (o más) que el que simplemente hubieras tenido un error en la documentación? ¿Poniendo un fix a un issue *que ponía explícitamente la línea donde estaba el error*?

E, insisto, ¿entiendes que todo eso habría pasado (ish, acabo de ver que no hiciste caso al issue de la forma como lo planteé, pero igual no lo hubiera visto) si no hubieras cometido el error bastante grave de poner el framework como dependencia de desarrollo?

Balrrach commented:
Entendido.
